### DISCUSSION TITLE
Creating local media tracks from frame queue

### DISCUSSION THREAD

**Discussion by keremcc** [aiortc/aiortc#607] [Q&A]
Hi, I was able to run aiortc webcam example. It works as expected if we want to use webcam or any usb camera directly.
My question: 
Is there a way to create localmediatracks from continuous frame queue that is captured with camera sdk to stream it to browser? 
Or more proper way is to use virtual camera?
Camera that I use is intel-realsense.
Any help is appreciated.

**Comment by danielkoohmarey-bc** [comment]
You could use something like:

```
  class ImageRenderingTrack(VideoStreamTrack):
  
      def __init__(self):
          super().__init__()
          self.queue = asyncio.Queue(10)
  
      async def add_image(self, img: numpy.ndarray):
          await self.queue.put(img)
  
      async def recv(self):
          img = await self.queue.get()
          frame = VideoFrame.from_ndarray(img, format="bgr24")
          pts, time_base = await self.next_timestamp()
          frame.pts = pts
          frame.time_base = time_base
          return frame
```
and call `add_image` from a separate thread /callback that receives the images (raw array) from your camera. The `ImageRenderingTrack` instance can then be added to the peer connection with `addTrack`

**Reply by ibrandiay** [reply]
@danielkoohmarey-bc excellent, thank you very much

**Comment by prasunnyD** [comment]
@danielkoohmarey-bc could you provide an example on how you would call add_image from a separate thread/callback? Thank you

**Reply by jlaine** [reply]
Hi! I've converted this issue to a discussion which seems like a better format for community interactions. I'm having a hard type keeping on top of issues!

**Reply by DanielKoohmarey** [reply]
@prasunnyD I've written a [small blog post with the details](https://pericror.com/software/python-create-a-webrtc-video-stream-from-images/)!
